File: leetcode/14D Algo2/BFS_DFS/117. Populating Next Right Pointers in Each Node II.java
import java.util.*;

// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};

/*
Populate each next pointer to point to its next right node. 
If there is no next right node, the next pointer should be set to NULL.
*/

class Solution {


    public Node connect(Node root){
        if(root == null) return root;
        Queue<Node> Q = new LinkedList<>();
        Q.add(root);
        while(!Q.isEmpty()){
            int len = Q.size();
            for (int i = 0; i < len; i++) {
                Node t = Q.poll();
                if(i < len - 1){
                    t.next = Q.peek();
                }
                if(t.left != null){
                    Q.add(t.left);
                }
                if(t.right != null){
                    Q.add(t.right);
                }
            }
        }
        return root;
    }













    public Node connect_1(Node root) {
        if(root == null) return root;
        Queue<Node> Q = new LinkedList<>();
        Q.add(root);

        while(!Q.isEmpty()){
            int len = Q.size();
            
            for(int i = 0; i < len; ++i){
                Node p = Q.poll();
                if(i != len - 1){ // in case Node is end of current level
                    p.next = Q.peek();
                }
                if(p.left != null) Q.add(p.left);
                if(p.right != null) Q.add(p.right);
            }

        }
        return root;
    }

}

File: leetcode/14D Algo2/BFS_DFS/117. Populating Next Right Pointers in Each Node II.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<list>
#include<stack>
#include<unordered_set> 
#include<unordered_map> 
using namespace std;

/*
Populate each next pointer to point to its next right node. 
If there is no next right node, the next pointer should be set to NULL.

Initially, all next pointers are set to NULL.
*/


// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};

class Solution {
public:
    Node* connect(Node* root) {
        if(root == nullptr) return root;
        queue<Node*> Q;
        Q.push(root);
        while(!Q.empty()){
            int count = Q.size();
                  
            for(int i = 0; i < count; ++i){
                Node* p = Q.front(); // extract 1 node
                Q.pop();
                
                // if not the last node in this level
                if(i != count - 1){ 
                    p->next = Q.front();
                }
                
                if(p->left) Q.push(p->left); // push left child
                if(p->right) Q.push(p->right); // push right child
            }
        }
        return root;
    }
};
